### ISSUE TITLE
Support SnsMessageManager

### ISSUE THREAD

**Issue by zoewangg** [aws/aws-sdk-java-v2#1302]
This issue tracks the 2.x implementation of equivalent functionality to the 1.11.x `SnsMessageManager`

**Comment by henricook** [comment]
I wanted to use a library to validate incoming SNS messages (notifications received from an SES ConfigurationSet) and had to use v1 because of the lack of Message Manager. As i'm on a Scala project after validating i just parse it off into a case class and don't really make use of the rest of the library.

If it sounds like i've done the right thing for now, great and :+1: for future V2 integration of a validator of some sort. If you think i could have done the validation with V2 please do let me know!

**Comment by Saberos** [comment]
Any chance to also resolve this issue in the SDK v2 implementation: https://github.com/aws/aws-sdk-net/issues/1104?

Just ran into the same issue in the Java SDK v1 implementation. SNS->Lambda contains field `SignatureCertUrl`, while `SignatureChecker` expects `SignatureCertURL`.

**Comment by lpellegr** [comment]
3 years after the initial report there is still no equivalent for a basic security check that I guess most SNS users should make...

**Comment by ryanthon** [comment]
Is there any update on implementing this?

**Comment by henricook** [comment]
Three years on and I'm still using V1 as in my original comment. If you want to provide a migration path off using SnsMessageManager instead of replacing it in code that would also be helpful

**Comment by kazaff** [comment]
Can't believable this problom is still existing.
How can I verify the sns message by spring-cloud-aws-messaging?
Does anyone solve this issue?

**Comment by kazaff** [comment]
> Three years on and I'm still using V1 as in my original comment. If you want to provide a migration path off using SnsMessageManager instead of migrating it to V2 in code that could also be helpful

Hi, I wonder that how you can still using V1?  I mean the SNS service had been updated, but you can still use Public Key method to verfiy message? I have try to create a new topic and found that no where to get the public key ...

**Comment by henricook** [comment]
```scala
  private val messageManager = new SnsMessageManager(Regions.EU_WEST_1.getName)

  def parseMessage(messageBody: InputStream): Try[Unit] = Try(messageManager.parseMessage(messageBody))
```

Takes care of everything for you 🤷🏻  - this is what you can't do in V2 (it seems) currently

**Comment by kazaff** [comment]
> ```scala
>   private val messageManager = new SnsMessageManager(Regions.EU_WEST_1.getName)
> 
>   def parseMessage(messageBody: InputStream): Try[Unit] = Try(messageManager.parseMessage(messageBody))
> ```
> 
> Takes care of everything for you 🤷🏻 - this is what you can't do in V2 (it seems) currently

Sorry, I can't understand your code... it's scala?
I want make sure that you mixed use V1 (only for verification) and V2 (for rest of stuff) in your project?
As far as I know, V1 verifies the message by Public Key, but this method don't support V2(at least I can't find the Public Key signature or something like that on V2 documents)..

**Comment by henricook** [comment]
No, i'm only using V1 in my project. I'm blocked from using V2 because I want the ability to verify received messages from SES/SNS callbacks using the library, like I do in V1 with `SnsMessageManager`.

**Comment by kazaff** [comment]
> No, i'm only using V1 in my project. I'm blocked from using V2 because I want the ability to verify received messages from SES/SNS-JSON-HTTP callbacks using the library, like I do in V1 with `SnsMessageManager`.

oh.. Got it. Thank you for your help, my bro.

**Comment by kazaff** [comment]
I have push some code to spring-cloud-aws(https://github.com/awspring/spring-cloud-aws/pull/684) which base on V2.
It can verify the message by sns signingCertUrl.

Hope will help you guys.

**Comment by sindhu-cbs** [comment]
Hi All, any update on this issue? I'm using v1 for parseMessage() in SnsMessageManager.java. Is there an equivalent method in v2 to validate the authenticity of an incoming SNS message?

**Comment by StephanWels** [comment]
@sindhu-cbs - same here! Can't find the equivalent in v2. Guess we haver to stick to v1 for now?

**Comment by henricook** [comment]
Missing `parseMessage` is the only reason we've been on V1 for the last two years+

**Comment by lpellegr** [comment]
As of July 31, 2024 v1 is deprecated but still not migration path for `SnsMessageManager::parseMessage`...

**Comment by henricook** [comment]
Is there some other way to parse a message/check its signature that I don't know about? It seems like such a bizarre gap 😅

**Comment by pbartoszek** [comment]
I am shocked it's not in v2 too...

**Comment by detinho** [comment]
> Is there some other way to parse a message/check its signature that I don't know about? It seems like such a bizarre gap 😅

@henricook you have to do it manually as my current understanding. I seems to me that they are pushing to use SQS or other AWS services as a target endpoint.

**Comment by efenderbosch-atg** [comment]
When this gets implemented, please support signature verification when using LocalStack for local development and testing. The current verification enforces https.
Example `SigningCertURL` for LocalStack: `http://localhost.localstack.cloud:4566/_aws/sns/SimpleNotificationService-6c6f63616c737461636b69736e696365.pem`

**Comment by lpellegr** [comment]
Since this issue has persisted since 2019, it seems unlikely it will ever be resolved or implemented. It would be highly appreciated to have an official response outlining the recommended alternative.

**Comment by tmccombs** [comment]
v1 will be EOL at the end of this year. Any replacement in sight?

**Comment by mikeleegdit** [comment]
v1 is now EOL. Is there any recommended workaround or replacement?

**Comment by henricook** [comment]
I just extracted the appropriate parts of the v1 source and wrote my own tbh

**Comment by dagnir** [comment]
Hi all, we are close to releasing the initial version of the message manager for 2.x: https://github.com/aws/aws-sdk-java-v2/pull/6804. Thank you for your patience and interest!

**Comment by dagnir** [comment]
Hi all! We have released this library as part of [2.42.35](https://github.com/aws/aws-sdk-java-v2/blob/2.42.35/CHANGELOG.md#amazon-sns-message-manager).
